tests/test_playbooks/job_invocation.yml

```diff
@@ -0,0 +1,34 @@
+# registered host expected
+
+---
+- hosts: tests
+  collections:
+    - theforeman.foreman
+  gather_facts: false
+  vars_files:
+    - vars/server.yml
+  vars:
+  tasks:
+    - name: "Run remote command on a single host once"
+      job_invocation:
+        username: "{{ foreman_username }}"
+        password: "{{ foreman_password }}"
+        server_url: "{{ foreman_server_url }}"
+        search_query: "{{ foreman_host }}"
+        command: 'ls'
+        job_template: 'Run Command - SSH Default'
+        validate_certs: "{{ foreman_validate_certs }}"
+
+    - name: "Run Ansible command on active hosts once a day"
```

Comment:
our tests usually follow the following pattern:
1. create an entity (ensure `changed=True`)
2. invoke the module with identical parameters, as the entity already exists, the module should return `changed=False`
3. update the entity (ensure `changed=True)
4. invoke the module with identical parameters, as the entity is already updated, the module should return `changed=False`
5. delete the entity (ensure changed=True)
6. invoke the module with identical parameters, as the entity is already deleted, the module should return `changed=False`

is that something that can be done with job invocations too?

Reply:
Since the entity is job invocation itself, I'd say most of the points don't apply. It is perfectly ok to invoke the job with same parameters, and once it is completed, there's no way to update or delete the record of it. 

For future-scheduled jobs, abortion is possible, I'll look into how to test it.

Recurring jobs can be updated and deleted, though I suppose as a part of the recurring job endpoint we didn't implement yet.

Reply:
you have `recurrence` param, isn't that what creates a recurring job?

Reply:
Yes that is how you create it, but if you want to revisit it, it's called `recurring logics` in the api. This recurring logic will trigger job invocations in the future. If you list job invocations, you can see those that already run and the next one queued by the rec. logic. So you can cancel the next planned job, but to update or cancel all the future repetitions you need to talk to recurring logics (seems that UI only supports cancelling logics, API can also update them)

Reply:
Oh Chtulhu. It goes via https://github.com/theforeman/foreman-tasks/blob/master/app/controllers/foreman_tasks/api/recurring_logics_controller.rb? "Nice"

The I think this module should not have state (absent) support at all for now, as it only can create invocations.

Reply:
Should it then change to be imperative (invoke_job) instead of declarative (job_invocation) to mark it fit for use in a handler without sacrificing idempotence of your playbook?

Reply:
@mdellweg  what would this entail? Is it just about renaming the file?

Reply:
And setting the right name in the documentation metadata, but yeah, that's all that's needed.

FWIW, I am fine with the current name (if you need another opinion 😅).

Just for completeness, what's the UI element called where you can schedule jobs?

Reply:
Heh, to run job inv in UI you usually go through the Hosts page (select host(s), pick Schedule Remote Job); or you can go through Job Templates (click Run). To list invocations you go to Job Invocations page (go to Monitor > Jobs) -- from there you can run jobs as well. Maybe there are other starting points I don't know of yet :)